services/staging/rs_server_staging/processors.py

```diff
@@ -11,22 +11,950 @@
+            contain only those that have not been downloaded yet.
+
+        Side Effects:
+            - Updates `self.stream_list` with the features that still need to be downloaded.
+
+        """
+        # Based on catalog response, pop out features already in catalog and prepare rest for download
+        try:
+            if not catalog_response["features"]:
+                # No search result found, process everything from self.item_collection
+                self.stream_list = self.item_collection.features
+            else:
+                # Do the difference, call rs-server-download only with features to be downloaded
+                # Extract IDs from the catalog response directly
+                already_downloaded_ids = {feature["id"] for feature in catalog_response["features"]}
+                # Select only features whose IDs have not already been downloaded (returned in /search)
+                not_downloaded_features = [
+                    item for item in self.item_collection.features if item.id not in already_downloaded_ids
+                ]
+                self.stream_list = not_downloaded_features
+        except KeyError as ke:
+            self.logger.exception(
+                "The 'features' field is missing in the response from the catalog service. "
+                f"Unable to check the collection {self.catalog_collection}. {ke}",
+            )
+            raise RuntimeError(
+                "The 'features' field is missing in the response from the catalog service. ",
+            ) from ke
+
+    def prepare_streaming_tasks(self, feature):
+        """Prepare tasks for the given feature to the Dask cluster.
+
+        Args:
+            feature: The feature containing assets to download.
+
+        Returns:
+            True if the info has been constructed, False otherwise
+        """
+
+        for asset_name, asset_content in feature.assets.items():
+            if not asset_content.href or not asset_content.title:
+                self.logger.error("Missing href or title in asset dictionary")
+                return False
+            # TODO: add the user_collection as main directory
+            s3_obj_path = f"{self.catalog_collection}/{feature.id.rstrip('/')}/{asset_content.title}"
+            self.assets_info.append((asset_content.href, s3_obj_path))
+            # update the s3 path, this will be checked in the rs-server-catalog in the
+            # publishing phase
+            asset_content.href = f"s3://rtmpop/{s3_obj_path}"
+            feature.assets[asset_name] = asset_content
+        return True
+
+    def handle_task_failure(self, error):
+        """Handle failures during task processing, including cancelling tasks and cleaning up S3 objects.
+
+        Args:
+            error (Exception): The exception that occurred.
+            tasks (list): List of Dask task futures.
+            s3_objs (list): List of S3 object paths to clean up.
```

Comment:
this is not an argument

Reply:
I forgot to delete them, from an older implementation. Done